File: HW/HW9/hw9.py
class TreeNode(object):
     def __init__(self, val=0, left=None, right=None):
        #python definiiton for treenodes
         self.val = val
         self.left = left
         self.right = right

class hw9:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not root:
            return None #edge case if it doesnt exist

        if root == p or root == q:
            return root #if the root is equal to p or q, return it , root can be LCA of itself, returns itself
        
        #  look in left and right child if those cases are not true
        left = right = None #initialiazing to nothing first
        if root.left:
            left = self.lowestCommonAncestor(root.left, p, q) #recursively searching the left intil a vaue is foumd
        if root.right:
            right = self.lowestCommonAncestor(root.right, p, q)

        # if both children returned a node, it means both p and q have been found so its parent is the LCA
        if left and right:
            return root
        else:
        # otherwise, only one of the chidren returned a node, meaning either p or q ertr found on left or right branch. 
        # in this case where the node was found would be the LCA
            return left or right
        

        

# Create the tree
tree1 = 
#given test cases
print(hw9().lowestCommonAncestor(tree1, 1, 5)) #4, 4 is the only common ancestor of 1 and 5.
print(hw9().lowestCommonAncestor(tree1, 3, 1)) # expected output 3,  Since 3 is the parent of 1, 3 is the lowest common ancestor of 3 and 1.  

File: HW/HW9/hw9.java
package HW.HW9;

public class hw9 {
    public class TreeNode { //java definition for treenodes
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {}
   
        TreeNode(int val) { this.val = val; }
   
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
     }

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null) //if the root is null we cant find the lca
            return null;
        
        if(root == p || root == q) //returns root value if equal to p or q, root can be LCA of itself
            return root;

        TreeNode left = lowestCommonAncestor(root.left, p, q); //recursively searches the left subtree, finding p or q or null
        TreeNode right = lowestCommonAncestor(root.right, p, q); //"" but for right subtree

         if(left == null) { //if we didn't find any value on the left (it is null), then we return what we have, and it would be the LCA
            return right;
        }
        else if(right == null) {
            return left;
        }
        else { //both left and right are not null, we found our result
            return root;
        }
    }

    public static void main(String[] args) {
        hw9 tree = new hw9();

        // creating binary tree
        TreeNode root = tree.new TreeNode(4);
        TreeNode n3 = tree.new TreeNode(3);
        TreeNode n8 = tree.new TreeNode(8);
        TreeNode n1 = tree.new TreeNode(1);
        TreeNode n5 = tree.new TreeNode(5);
        TreeNode n9 = tree.new TreeNode(9);

        root.left = n3;
        root.right = n8;
        n3.left = n1;
        n8.left = n5;
        n8.right = n9;

        // 1st test
        TreeNode p = n1;
        TreeNode q = n5;
        TreeNode lca = tree.lowestCommonAncestor(root, p, q);
        System.out.println("lca of 1 and 5: "+ lca.val); // Expected: 4

        // 2nd test 
        p = n3;
        q = n1;
        lca = tree.lowestCommonAncestor(root, p, q);
        System.out.println("lca of 1 and 3: "+ lca.val); // Expected: 3

    }
}
